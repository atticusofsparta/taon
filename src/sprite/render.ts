export const animationNames = ['no_anim', 'idle', 'run', 'jump'] as const;

const ANIMATIONS: Record<typeof animationNames[number], { len: number, y: number }> = {
	no_anim: { len: 1, y: 0 },
	idle: { len: 4, y: 1 },
	run: { len: 4, y: 2 },
	jump: { len: 7, y: 3 },
};

type Pos = {
	x: number,
	y: number,
}

type Size = {
	w: number,
	h: number,
}

type Rect = Pos & Size;

type Meta = {
	image: string,
	format: string,
	size: Size,
	scale: number,
}

type Frame = {
	frame: Rect,
	sourceSize: Size,
	spriteSourceSize: Rect,
}

type FrameName = string

type Frames = Record<FrameName, Frame>

type Animations = Record<string, string[]>

export type SpriteData = {
	meta: Meta,
	frames: Frames,
	animations: Animations,
}

const spriteSize: Size = { w: 16, h: 24 };

function generateFrameNames(prefix: string, count: number): FrameName[] {
	return Array.from(Array(count).keys()).map((i) => prefix + i.toString());
}

function generateFrames(frameNames: string[], yOffset: number): Frames {
	return frameNames.map((e, i) => ({
		[e]: {
			frame: {
				x: i * spriteSize.w,
				y: yOffset * spriteSize.h,
				w: spriteSize.w,
				h: spriteSize.h,
			},
			sourceSize: spriteSize,
			spriteSourceSize: {
				x: 0,
				y: 0,
				w: spriteSize.w,
				h: spriteSize.h,
			}
		}})).reduce((a, b) => ({...a,...b}), {});
}

const animations = animationNames.map((e) => {
	const anim = ANIMATIONS[e];
	const frameNames = generateFrameNames(e, anim.len);
	return {
		[e]: frameNames,
	};
}).reduce((a, b) => ({...a,...b}), {});

const frames = animationNames
	.map((e, i) => generateFrames(animations[e], i))
	.reduce((a, b) => ({...a,...b}), {});

export function generateSpriteData(spriteSheetUrl: string): SpriteData {
	return {
		meta: {
			image: spriteSheetUrl,
			format: "RGBA8888",
			size: spriteSize,
			scale: 1,
		},
		frames,
		animations,
	};
}
