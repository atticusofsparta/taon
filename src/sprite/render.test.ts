import { describe, it, expect } from "bun:test";
import { generateSpriteData } from "./render";

describe("generateSpriteData", () => {
	it("should generate sprite data", () => {
		const data = generateSpriteData("/api/sprite/generate?seed=123");
		expect(data).toMatchSnapshot();
	});
});
