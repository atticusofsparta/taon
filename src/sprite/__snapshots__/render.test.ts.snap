// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateSpriteData > should generate sprite data 1`] = `
{
  "animations": {
    "idle": [
      "idle0",
      "idle1",
      "idle2",
      "idle3",
    ],
    "jump": [
      "jump0",
      "jump1",
      "jump2",
      "jump3",
      "jump4",
      "jump5",
      "jump6",
    ],
    "no_anim": [
      "no_anim0",
    ],
    "run": [
      "run0",
      "run1",
      "run2",
      "run3",
    ],
  },
  "frames": {
    "idle0": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 24,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "idle1": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 16,
        "y": 24,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "idle2": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 32,
        "y": 24,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "idle3": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 48,
        "y": 24,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "jump0": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 72,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "jump1": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 16,
        "y": 72,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "jump2": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 32,
        "y": 72,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "jump3": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 48,
        "y": 72,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "jump4": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 64,
        "y": 72,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "jump5": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 80,
        "y": 72,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "jump6": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 96,
        "y": 72,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "no_anim0": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "run0": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 48,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "run1": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 16,
        "y": 48,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "run2": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 32,
        "y": 48,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "run3": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 48,
        "y": 48,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
  },
  "meta": {
    "format": "RGBA8888",
    "image": "/api/sprite/generate?seed=123",
    "scale": 1,
    "size": {
      "h": 24,
      "w": 16,
    },
  },
}
`;

exports[`generateSpriteData should generate sprite data 1`] = `
{
  "animations": {
    "idle": [
      "idle0",
      "idle1",
      "idle2",
      "idle3",
    ],
    "jump": [
      "jump0",
      "jump1",
      "jump2",
      "jump3",
      "jump4",
      "jump5",
      "jump6",
    ],
    "no_anim": [
      "no_anim0",
    ],
    "run": [
      "run0",
      "run1",
      "run2",
      "run3",
    ],
  },
  "frames": {
    "idle0": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 24,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "idle1": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 16,
        "y": 24,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "idle2": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 32,
        "y": 24,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "idle3": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 48,
        "y": 24,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "jump0": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 72,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "jump1": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 16,
        "y": 72,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "jump2": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 32,
        "y": 72,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "jump3": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 48,
        "y": 72,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "jump4": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 64,
        "y": 72,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "jump5": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 80,
        "y": 72,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "jump6": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 96,
        "y": 72,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "no_anim0": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "run0": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 48,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "run1": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 16,
        "y": 48,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "run2": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 32,
        "y": 48,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
    "run3": {
      "frame": {
        "h": 24,
        "w": 16,
        "x": 48,
        "y": 48,
      },
      "sourceSize": {
        "h": 24,
        "w": 16,
      },
      "spriteSourceSize": {
        "h": 24,
        "w": 16,
        "x": 0,
        "y": 0,
      },
    },
  },
  "meta": {
    "format": "RGBA8888",
    "image": "/api/sprite/generate?seed=123",
    "scale": 1,
    "size": {
      "h": 24,
      "w": 16,
    },
  },
}
`;
